autonomous/autonomous.py

```diff
@@ -1,2 +1,144 @@
-class AutoBase:
-    ...
+from typing import Optional
+from magicbot.state_machine import AutonomousStateMachine, state
+from dataclasses import dataclass
+
+from components.chassis import Chassis
+from components.intake import Intake
+from components.arm import Arm, Setpoints, Setpoint
+from components.gripper import Gripper
+
+from controllers.movement import Movement
+
+from wpilib import Timer
+from wpimath.geometry import Pose2d, Rotation2d, Translation2d
+
+from utilities.functions import grid_col_to_field_y
+
+from math import radians
+
+
+@dataclass
+class PickupPath:
+    goal: Pose2d
+    approach_direction: Rotation2d
+    intermediate_waypoints: Optional[list[Translation2d]]
+
+
+@dataclass
+class ScorePath:
+    goal: Pose2d
+    approach_direction: Rotation2d
+    intermediate_waypoints: Optional[list[Translation2d]]
+    arm_setpoint: Setpoint
+
+
+class AutoBase(AutonomousStateMachine):
+    gripper: Gripper
+    arm: Arm
+    chassis: Chassis
+    intake: Intake
+    movement: Movement
+
+    def __init__(self) -> None:
+        self.piece_positions = []
+        self.pickup_paths = []
+        self.score_paths = []
+        self.stage = 0
+        self.timeout_set = False
+        self.timeout_start = 0
+
+    @state(first=True)
+    def score(self, state_tm: float, initial_call: bool) -> None:
+        if initial_call:
+            self.timeout_set = False
+            path = self.score_paths[self.stage]
+            self.movement.next_state("autodrive")
+            self.movement.set_goal(
+                path.goal,
+                path.approach_direction,
+                waypoints=path.intermediate_waypoints,
+            )
+        self.movement.do_autodrive()
+        if self.movement.time_remaining < 0.5:
+            # self.arm.set_setpoint(self.score_paths[self.stage].arm_setpoint)
+            pass
+        if (
+            self.movement.is_at_goal()
+            # and self.arm.at_goal_angle()
+            # and self.arm.at_goal_extension()
+        ):
+            self.gripper.open()
+        if self.gripper.opened:
+            # self.arm.set_setpoint(Setpoints.HANDOFF)
+            if self.stage >= len(self.pickup_paths):
+                self.done()
+            self.next_state("pickup_cube")
+
+    @state
+    def pickup_cube(self, state_tm: float, initial_call: bool) -> None:
+        if initial_call:
+            self.timeout_set = False
+            path = self.pickup_paths[self.stage]
+            self.movement.next_state("autodrive")
```

Comment:
It may be nicer to make set_goal call generate trajectory in movement, I think that would also fix time_remaining not being set correctly on the first run